cpmpy/transformations/decompose_global.py

```diff
@@ -1,107 +1,134 @@
+            newlist.append(expr)
+
+        elif isinstance(expr, GlobalConstraint):
+            expr = copy.copy(expr)
+            expr.args = decompose_in_tree(expr.args, supported, supported_nested, _toplevel, nested=True)
+
+            if expr.is_bool():
+                # boolean global constraints
+                is_supported = (nested and expr.name in supported_nested) or (not nested and expr.name in supported)
+                if not is_supported:
+                    if nested is False or expr.equivalent_decomposition():
+                        expr = decompose_in_tree(expr.decompose(), supported, supported_nested, _toplevel, nested=True)
+                        newlist.append(all(expr))  # can create top-level AND
+                    else:
+                        bv = boolvar()
+                        newlist.append(bv)
+                        impl = decompose_in_tree([bv.implies(d) for d in expr.decompose()], supported, supported_nested,
+                                                 _toplevel, nested=True)
+                        neg_impl = decompose_in_tree([(~bv).implies(d) for d in expr.decompose_negation()], supported,
+                                                     supported_nested, _toplevel, nested=True)
+                        _toplevel.extend([all(impl), all(neg_impl)])
+                else:
+                    newlist.append(expr)
+            else:
+                # numerical global function (e.g. min, max, count, element...) in non-comparison context
+                if (nested and expr.name not in supported_nested) or (not nested and expr.name not in supported):
+                    aux = intvar(*expr.get_bounds())
+                    newlist.append(aux)
+                    l1, l2 = expr.decompose_comparison("==", aux)
+                    newlist.extend(l1 + l2) # all constraints should be added toplevel, node replaced by aux var
+                else:
+                    newlist.append(expr)
+
+        elif isinstance(expr, Comparison):
+            # Tricky case, if we just recurse into arguments, we risk creating unnecessary auxiliary variables
+            #  e.g., min(x,y,z) == a would become `min(x,y,z).decompose_comparison('==',aux) + [aux == a]`
+            #   while more optimally, its just `min(x,y,z).decompose_comparison('==', a)`
+            #    so have to be careful here and operate from 'one level above'
+
+            lhs, rhs = expr.args
+            if hasattr(lhs, "decompose_comparison"):
+                lhs = copy.copy(lhs)
+                lhs.args = decompose_in_tree(lhs.args, supported, supported_nested, _toplevel, nested=True)
+                lhs_supported = (nested and lhs.name in supported_nested) or (not nested and lhs.name in supported)
+                if not lhs_supported:
+                    newexpr, _toplevel[len(_toplevel):] = lhs.decompose_comparison(expr.name, rhs)
```

Comment:
the in-place iterator trick is very subtle, too few people will understand it, please separate it out as a separate `.extend(l2)`